src/shared/image-policy.c

```diff
@@ -0,0 +1,647 @@
+         * return the default policy. */
+
+        flags = image_policy_get(policy, designator);
+        if (flags < 0)
+                return partition_policy_normalized_flags(
+                                &(const PartitionPolicy) {
+                                        .flags = image_policy_default(policy),
+                                        .designator = designator,
+                                });
+
+        return flags;
+}
+
+static PartitionPolicyFlags policy_flag_from_string_one(const char *s) {
+        assert(s);
+
+        /* This is a bitmask (i.e. not dense), hence we don't use the "string-table.h" stuff here. */
+
+        if (streq(s, "verity"))
+                return PARTITION_POLICY_VERITY;
+        if (streq(s, "signed"))
+                return PARTITION_POLICY_SIGNED;
+        if (streq(s, "encrypted"))
+                return PARTITION_POLICY_ENCRYPTED;
+        if (streq(s, "unprotected"))
+                return PARTITION_POLICY_UNPROTECTED;
+        if (streq(s, "unused"))
+                return PARTITION_POLICY_UNUSED;
+        if (streq(s, "absent"))
+                return PARTITION_POLICY_ABSENT;
+        if (streq(s, "open")) /* shortcut alias */
+                return PARTITION_POLICY_OPEN;
+        if (streq(s, "ignore")) /* ditto */
+                return PARTITION_POLICY_IGNORE;
+        if (streq(s, "read-only-on"))
+                return PARTITION_POLICY_READ_ONLY_ON;
+        if (streq(s, "read-only-off"))
+                return PARTITION_POLICY_READ_ONLY_OFF;
+        if (streq(s, "growfs-on"))
+                return PARTITION_POLICY_GROWFS_ON;
+        if (streq(s, "growfs-off"))
+                return PARTITION_POLICY_GROWFS_OFF;
+
+        return _PARTITION_POLICY_FLAGS_INVALID;
+}
+
+PartitionPolicyFlags partition_policy_flags_from_string(const char *s) {
+        PartitionPolicyFlags flags = 0;
+        int r;
+
+        assert(s);
+
+        if (streq(s, "-"))
+                return 0;
+
+        for (;;) {
+                _cleanup_free_ char *f = NULL;
+                PartitionPolicyFlags ff;
+
+                r = extract_first_word(&s, &f, "+", EXTRACT_DONT_COALESCE_SEPARATORS);
+                if (r < 0)
+                        return r;
+                if (r == 0)
+                        break;
+
+                ff = policy_flag_from_string_one(strstrip(f));
+                if (ff < 0)
+                        return -EBADRQC; /* recognizable error */
+
+                flags |= ff;
+        }
+
+        return flags;
+}
+
+static ImagePolicy* image_policy_new(size_t n_policies) {
+        ImagePolicy *p;
+
+        if (n_policies > (SIZE_MAX - offsetof(ImagePolicy, policies)) / sizeof(PartitionPolicy)) /* overflow check */
+                return NULL;
+
+        p = malloc(offsetof(ImagePolicy, policies) + sizeof(PartitionPolicy) * n_policies);
+        if (!p)
+                return NULL;
+
+        p->n_policies = 0;
```

Comment:
structured initialisation?

Reply:
It's a variable size structure, and compound initialization is really weird for that, as it only covers the fixed size part of it... I think we probably should not really make use of that too much, dunno.